File: go/blackjack/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
	case "two":
		return 2
	case "three":
		return 3
	case "four":
		return 4
	case "five":
		return 5
	case "six":
		return 6
	case "seven":
		return 7
	case "eight":
		return 8
	case "nine":
		return 9
	case "ten", "jack", "queen", "king":
		return 10
	case "ace":
		return 11
	default:
		return 0
	}
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	const (
		Stand = "S"
		Hit   = "H"
		Split = "P"
		Win   = "W"
	)

	val1, val2, dealerVal := ParseCard(card1), ParseCard(card2), ParseCard(dealerCard)
	sum := val1 + val2

	switch {
	// Split
	case val1 == val2 && val1 == 11:
		return Split

	// Win
	case sum == 21 && dealerVal < 10:
		return Win

	// Stand
	case sum == 21,
		17 <= sum && sum <= 20,
		12 <= sum && sum <= 16 && dealerVal < 7:
		return Stand

	// Hit
	default:
		return Hit
	}
}

File: ruby/blackjack/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
    when "ace" then 11
    when "jack", "queen", "king", "ten" then 10
    when "nine" then 9
    when "eight" then 8
    when "seven" then 7
    when "six" then 6
    when "five" then 5
    when "four" then 4
    when "three" then 3
    when "two" then 2
    when "one" then 1
    else 0
    end
  end

  def self.card_range(card1, card2)
  	case parse_card(card1) + parse_card(card2)
   	when 4..11 then "low"
    when 12..16 then "mid"
    when 17..20 then "high"
    when 21 then "blackjack"
    end
  end

  def self.first_turn(card1, card2, dealer_card)
  	range = card_range(card1, card2)
   	dealer_val = parse_card(dealer_card)
    case
    when card1 == "ace" && card2 == "ace" then "P"
    when range == "blackjack" && dealer_val < 10 then "W"
    when range == "blackjack", range == "high", range == "mid" && dealer_val < 7 then "S"
    when range == "mid", range == "low" then "H"
		end
  end
end
